lib/spack/spack/environment.py

```diff
@@ -1509,6 +1509,62 @@ def concretized_specs(self):
         for s, h in zip(self.concretized_user_specs, self.concretized_order):
             yield (s, self.specs_by_hash[h])
 
+    def matching_spec(self, spec):
+        """
+        Given a spec (likely not concretized), find a matching concretized
+        spec in the environment.
+
+        The matching spec does not have to be installed in the environment,
+        but must be concrete (specs added with `spack add` without an
+        intervening `spack concretize` will not be matched).
+
+        If there is a single root spec that matches the provided spec or a
+        single dependency spec that matches the provided spec, then the
+        concretized instance of that spec will be returned.
+
+        If multiple root specs match the provided spec, or no root specs match
+        and multiple dependency specs match, then this raises an error
+        and reports all matching specs.
+        """
+        # These are pairs of abstract/concrete specs (the abstract specs are
+        # useful for constructing error messages when there are multiple
+        # matches)
+        root_matches = list()
+        dep_matches = list()  # These are concrete
+        # These are concrete (for root specs, they store the concretized spec
+        # rather than the abstract spec)
+        all_matches = list()
+        for user_spec, concretized_user_spec in self.concretized_specs():
+            if concretized_user_spec.satisfies(spec):
+                root_matches.append((user_spec, concretized_user_spec))
+                all_matches.append(concretized_user_spec)
+            for dep_spec in concretized_user_spec.traverse(root=False):
+                if dep_spec.satisfies(spec):
+                    dep_matches.append(dep_spec)
```

Comment:
This will cause a spec to be repeated multiple times if it appears in multiple places. I'm trying to think through the cleanest way to resolve it.